rfcs/0004-pablo-distribution.adoc

```diff
@@ -0,0 +1,626 @@
+
+==== Staking
+
+[plantuml,images/staking,png]
+----
+start
+->assetId,\namount(in LP/PBLO/PICA);
+:rewards_pool = readStorage(assetId);
+:boosted_amount = reward_multiplier * amount;
+note right: Handle boosting as per reward_multiplier
+:new_pool_shares = boosted_amount;
+:new_position = StakingNFT{
+asset: assetId,
+share: boosted_amount,
+stake: amount,
+// ....
+};
+while (reward = rewards_pool.rewards.has_next) is (yes)
+    :inflation = reward.total_rewards * new_pool_shares /\nrewards_pool.total_shares;
+    :reward.total_rewards += inflation;
+    :reward.total_inflation += inflation;
+    :new_position.reductions[reward.asset_id] += inflation;
+endwhile (no)
+:rewards_pool.total_shares += boosted_amount;
+:save(new_position);
+:save(rewards_pool);
+stop
+----
+
+==== Extend Position
+
+[plantuml,images/extend-position,png]
+----
+start
+->positionId,assetId,\namount(in LP/PBLO/PICA)\ntime;
+:rewards_pool = readStorage(assetId);
+:boosted_amount = reward_multiplier * amount;
+note right: Handle boosting as per reward_multiplier
+:new_pool_shares = boosted_amount;
+:new_position = StakingNFT{
+asset: assetId,
+share: boosted_amount,
+stake: amount,
+// ....
+};
+while (reward = rewards_pool.rewards.has_next) is (yes)
+    :inflation = reward.total_rewards * new_pool_shares /\nrewards_pool.total_shares;
+    :reward.total_rewards += inflation;
+    :reward.total_inflation += inflation;
+    :new_position.reductions[reward.asset_id] += inflation;
+endwhile (no)
+:rewards_pool.total_shares += boosted_amount;
+:existing_position = readStorage(positionId);
+:merged_position = existing_position + new_position;
+note right: as per section 5.3.1.4
+:save(merged_position);
+:save(rewards_pool);
+stop
+----
+
+==== Split Position
+
+[plantuml,images/split-position,png]
+----
+start
+->assetId\npositionId\nratio;
+:rewards_pool = readStorage(assetId);
+:position = readStorage(positionId);
+:splits = split(position, ratio);
+note right: as per section 5.3.1.5
+:save(...splits);
+stop
+----
+
+==== Claim/Unstake
+
+[plantuml,images/claim,png]
+----
+start
+->assetId\npositionId;
+:rewards_pool = readStorage(assetId);
+:position = readStorage(positionId);
+while (reward = rewards_pool.rewards.has_next) is (yes)
+    :claim = reward.total_rewards * position.shares /\nrewards_pool.total_shares - position.reductions[reward.asset_id];
+    :reward.total_rewards -= position.reductions[reward.asset_id];
+    :reward.total_inflation -= position.reductions[reward.asset_id];
+    :transfer(claim);
+endwhile (no)
+:rewards_pool.total_shares -= position.shares;
+:remove(position);
+:save(rewards_pool);
+stop
+----
+
+==== Update Reward Pool
+
+[plantuml,images/update-reward-pool,png]
+----
+start
+->assetId\nreward_update\n;
+:rewards_pool = readStorage(assetId);
+while (reward = reward_update.rewards.has_next) is (yes)
+    :assert max_reward not exceeded;
+    :reward.total_rewards += reward_update[reward.asset_id].amount;
```

Comment:
Before updating the values, need to calculate the epoch rewards using previous values and add to the reward pool